#### Problem 1: (1) Solution

In [196]:
# Import Module
import os
import numpy as np
import pandas as pd
from numpy.linalg import norm



# Folder Path
path = '//home/sneha/Dropbox/Coursework_sem2/CS_633/Assignments/Assignment_1/docs'

# Change the directory
os.chdir(path)

vocabulary = set()

def create_word_vocabulary(file_path):

     data   = set(line.strip() for line in open(file_path, 'r'))
     vocabulary.update(data)

# iterate through all files
for file in os.listdir():
	 
     file_path = f"{path}/{file}"   
     create_word_vocabulary(file_path)

#Collecting all documents in docs
docs = os.listdir(path)
vocab1 = list(vocabulary)

vocab_dict = {}

for i, w in enumerate(vocab1):
     vocab_dict[w] = i


In [197]:
#Initial sparse-matrix creation for document vector
doc_vector = np.zeros((len(vocab1), len(docs)))

for file_i, file_name in enumerate(docs):
    file_path = f"{path}/{file_name}"   
    with open(file_path,'r') as f:
        for word in f.read().splitlines():
            doc_vector[vocab_dict[word]][file_i]+=1


In [198]:
#Collecting all queries into queries
query_path = "//home/sneha/Dropbox/Coursework_sem2/CS_633/Assignments/Assignment_1/queries"
queries = os.listdir(query_path)
queries.sort()

#Initial sparse-matrix creation for query vector
query_vector = np.zeros((len(vocab1), len(queries)))

for file_iq, file_nameq in enumerate(queries):
    file_path1 = f"{query_path}/{file_nameq}"   
    with open(file_path1,'r') as f:
        for word in f.read().splitlines():
            query_vector[vocab_dict[word]][file_iq]+=1


In [199]:
# Function to calculate dot product
def dotsimilarity(arr_doc, arr_qry):
    return np.dot(arr_doc, arr_qry)

# Function to calculate cosine similarity
def cossimilarity(arr_doc, arr_qry):
    return np.dot(arr_doc,arr_qry)/ (norm(arr_doc)*norm(arr_qry))

In [200]:
#Calculating similarity score for both:
doc_query_dot_score = np.zeros((len(docs), len(queries)))
doc_query_cosine_score = np.zeros((len(docs), len(queries)))

for query_ind, arr_qry in enumerate(query_vector.T):
 for doc_ind, arr_doc in enumerate(doc_vector.T):
    doc_query_dot_score[doc_ind][query_ind] = dotsimilarity(arr_doc, arr_qry)
    doc_query_cosine_score[doc_ind][query_ind] = cossimilarity(arr_doc, arr_qry)

In [246]:
#Indexes of top ten documents for dot similarity
doc_query_dot_score_tr = doc_query_dot_score.transpose()
top_docs_dot = np.empty((5,10))
for i in range(0, len(doc_query_dot_score_tr)):
    # top_docs[i] = np.argpartition(doc_query_dot_score_tr[i], -10)[-10:]
    temp = np.argpartition(doc_query_dot_score_tr[i], -10)[-10:]
    temp = list(temp[np.argsort(doc_query_dot_score_tr[i][temp])])
    temp.reverse()
    top_docs_dot[i] = temp.copy()
    indices_dot = [int(j) for j in top_docs_dot[i]]
    doc_query_dot_score_tr[i][indices_dot]
    print("The dot similarity scores for query {query_no}".format(query_no=i+1), doc_query_dot_score_tr[i][indices_dot])



The dot similarity scores for query 1 [170.  21.  11.  10.   9.   7.   7.   6.   6.   5.]
The dot similarity scores for query 2 [64. 43. 30. 19. 17. 15. 12. 11.  7.  6.]
The dot similarity scores for query 3 [9. 7. 5. 5. 4. 3. 3. 3. 3. 2.]
The dot similarity scores for query 4 [10.  6.  6.  4.  3.  3.  3.  3.  2.  2.]
The dot similarity scores for query 5 [23. 17.  7.  6.  6.  5.  5.  4.  4.  4.]


In [202]:
np.argpartition(doc_query_dot_score_tr[0], -10)[-10:]
# list(np.argpartition(doc_query_dot_score_tr[i], -10)[-10:][np.argsort(doc_query_dot_score_tr[i][10])])

array([154, 111, 435, 413, 340,  33, 446, 183,  97, 303])

In [247]:
#Top 10 documents for each of the five queries using dot similarity
top_docs_dot_index = np.empty((5,10))
for i in range(0, len(top_docs_dot)):
    for j in range(0,len(top_docs_dot[i])):
        top_docs_dot_index[i][j] = docs[int(top_docs_dot[i][j])]

print("Top ten documents in descending order for each query from 1-5 using dot similarity are:\n", top_docs_dot_index)

Top ten documents in descending order for each query from 1-5 using dot similarity are:
 [[ 51060.  49960.  51153.  51165.  51164.  51144.  51120.  51130.  51135.
   51161.]
 [ 59905.  60103.  59850.  59874.  59873.  59909.  59871.  59870.  60210.
   59913.]
 [ 10011.  59913.  59850.  10083.  10089.  59874.  59873.  10074.  10066.
   59848.]
 [ 59849.  59850.  51151.  51211.  59874.  59905.  60213.  60152.  60103.
   59908.]
 [102610. 101666. 102591. 102647. 102604. 102627. 102609. 102648. 102598.
  100521.]]


In [244]:
#Indexes of top ten documents for cosine similarity
doc_query_cosine_score_tr = doc_query_cosine_score.transpose()
top_docs_cos = np.empty((5,10))
for i in range(0, len(doc_query_cosine_score_tr)):
    # top_docs[i] = np.argpartition(doc_query_dot_score_tr[i], -10)[-10:]
    temp = np.argpartition(doc_query_cosine_score_tr[i], -10)[-10:]
    temp = list(temp[np.argsort(doc_query_cosine_score_tr[i][temp])])
    temp.reverse()
    top_docs_cos[i] = temp.copy()
    indices_cos = [int(j) for j in top_docs_cos[i]]
    doc_query_cosine_score_tr[i][indices_cos]
    print("The cosine similarity scores for query {query_no}".format(query_no=i+1), doc_query_cosine_score_tr[i][indices_cos])

  

The cosine similarity scores for query 1 [0.56104694 0.27498597 0.23981605 0.23054198 0.21650635 0.21004201
 0.20100756 0.19364917 0.17770466 0.17251639]
The cosine similarity scores for query 2 [0.26579595 0.25147489 0.22541741 0.219826   0.20988774 0.15075567
 0.14953715 0.13074409 0.1294745  0.12751534]
The cosine similarity scores for query 3 [0.2236068  0.17334381 0.17149859 0.13545709 0.13245324 0.12909944
 0.11338934 0.09097817 0.08111071 0.07624929]
The cosine similarity scores for query 4 [0.1118034  0.10540926 0.09759001 0.09325048 0.09263717 0.08679261
 0.08272911 0.07784989 0.06593805 0.06593805]
The cosine similarity scores for query 5 [0.39223227 0.37369979 0.2773501  0.2305734  0.22011273 0.20519567
 0.19245009 0.17342199 0.16439899 0.16081688]


In [248]:
#Top 10 documents for each of the five queries using cosine similarity
top_docs_cos_index = np.empty((5,10))
for i in range(0, len(top_docs_cos)):
    for j in range(0,len(top_docs_cos[i])):
        top_docs_cos_index[i][j] = docs[int(top_docs_cos[i][j])]
        
print("Top ten documents in descending order for each of queries 1-5 using cosine similarity are:\n", top_docs_cos_index)

Top ten documents in descending order for each of queries 1-5 using cosine similarity are:
 [[ 51060.  51144.  51120.  51164.  51135.  51158.  51184.  51171.  51130.
   51161.]
 [ 59905.  60103.  60170.  60210.  59850.  60195.  59909.  60198.  59870.
   60200.]
 [ 10064.  10083.  10063.  10089.  10052.  10013.  10066.  59913.  10067.
  101639.]
 [102656. 102660.  51207.  10027.  51151.  59849.  60213. 102675. 102626.
   51206.]
 [102598. 102610. 102647. 101666. 102609. 100521. 102617. 102608. 102634.
  102622.]]


#### Problem1: (2) Solution

In [218]:
df_dot = pd.DataFrame(top_docs_dot_index, columns = ['Rank1','Rank2','Rank3', 'Rank4', 'Rank5', 'Rank6', 'Rank7', 'Rank8', 'Rank9', 'Rank10'])
df_dot.index = ['Query1', 'Query2', 'Query3', 'Query4', 'query5']

In [216]:
df_dot

Rank1     Rank2     Rank3     Rank4     Rank5     Rank6     Rank7  \
Query1   51060.0   49960.0   51153.0   51165.0   51164.0   51144.0   51120.0   
Query2   59905.0   60103.0   59850.0   59874.0   59873.0   59909.0   59871.0   
Query3   10011.0   59913.0   59850.0   10083.0   10089.0   59874.0   59873.0   
Query4   59849.0   59850.0   51151.0   51211.0   59874.0   59905.0   60213.0   
query5  102610.0  101666.0  102591.0  102647.0  102604.0  102627.0  102609.0   

           Rank8     Rank9    Rank10  
Query1   51130.0   51135.0   51161.0  
Query2   59870.0   60210.0   59913.0  
Query3   10074.0   10066.0   59848.0  
Query4   60152.0   60103.0   59908.0  
query5  102648.0  102598.0  100521.0

In [219]:
df_cosine = pd.DataFrame(top_docs_cos_index, columns = ['Rank1','Rank2','Rank3', 'Rank4', 'Rank5', 'Rank6', 'Rank7', 'Rank8', 'Rank9', 'Rank10'])
df_cosine.index = ['Query1', 'Query2', 'Query3', 'Query4', 'query5']

In [220]:
df_cosine

Rank1     Rank2     Rank3     Rank4     Rank5     Rank6     Rank7  \
Query1   51060.0   51144.0   51120.0   51164.0   51135.0   51158.0   51184.0   
Query2   59905.0   60103.0   60170.0   60210.0   59850.0   60195.0   59909.0   
Query3   10064.0   10083.0   10063.0   10089.0   10052.0   10013.0   10066.0   
Query4  102656.0  102660.0   51207.0   10027.0   51151.0   59849.0   60213.0   
query5  102598.0  102610.0  102647.0  101666.0  102609.0  100521.0  102617.0   

           Rank8     Rank9    Rank10  
Query1   51171.0   51130.0   51161.0  
Query2   60198.0   59870.0   60200.0  
Query3   59913.0   10067.0  101639.0  
Query4  102675.0  102626.0   51206.0  
query5  102608.0  102634.0  102622.0

#### * Cosine similarity is considered a better metrics than dot similarity because it is more robust to changes in vector magnitude.<br>
#### * It is less affected by the magnitude of the vectors, which can make a big difference in similarity scores. <br>
#### * Additionally, cosine similarity is more easily interpretable since the scores range from 0 to 1, making it easier to understand how similar two vectors are whereas dot product can range between {-infinity, + infinity}. <br>
#### * Without normalization, there is a substancial chance that a bigger document return a higher similiarity score since it has larger number of words and hence it has a higher chance of having overlapping words with the query. <br>

#### * The above can be verified from below example. Consider *Query 1* - <br>
#### - Top 3 ranked documents using Dot Product : 51060, 49960, 51153 <br>
#### - Top 3 ranked documents using Cosine Product : 51060, 51144, 51120 <br>
#### First document is same for both, but for second and third, the documents are different. <br>

##### Following are the results for Query 1: <br>
No: of words in Query :  6<br>
Words in Query 1<br>
Word :  religious 	 Count :  3<br>
Word :  christian 	 Count :  1<br>
Word :  religion 	 Count :  1<br>
Word :  belief 	 Count :  1<br>

<br>
Dot Product Rank 2 Document : 49960<br>
Overlapping words in Document 49960<br>
Word :  religious 	 Count :  3<br>
Word :  christian 	 Count :  4<br>
Word :  religion 	 Count :  6<br>
Word :  belief 	 Count :  2<br>
No: of words in Document 49960 :  465<br>
Total no: of overlapping words in Document 49960 :  15<br>
<br>
<br>
Cosine Similarity Product Rank 2 Document : 51144<br>
Overlapping words in Document 51144<br>
Word :  religious 	 Count :  2<br>
Word :  christian 	 Count :  0<br>
Word :  religion 	 Count :  1<br>
Word :  belief 	 Count :  0<br>
No: of words in Document 51144 :  40<br>
Total no: of overlapping words in Document 51144 :  3<br>

#### It is observed that in Document 51144, (3 / 40) (7.5%) are overlapping words with respect to Query 1 but in Document 49960 there are only (15 / 465) words that are overlapping with Query 1. Hence, Document 51144 (returned by Cosine similarity metric) has a higher ratio of overlap with query explaining its better similarity over dot similarity.